mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java

```diff
@@ -346,14 +348,15 @@ public void close() throws IOException {
       DataFile file = currentTask.file();
       // schema of rows returned by readers
       PartitionSpec spec = currentTask.spec();
-      Set<Integer> idColumns = spec.identitySourceIds();
       Schema readSchema = expectedSchema != null ? expectedSchema : tableSchema;
+      Set<Integer> idColumns =  Sets.intersection(spec.identitySourceIds(), TypeUtil.getProjectedIds(readSchema));
       boolean hasJoinedPartitionColumns = !idColumns.isEmpty();
+
       if (hasJoinedPartitionColumns) {
-        readSchema = TypeUtil.selectNot(tableSchema, idColumns);
```

Comment:
Had to change this logic slightly so that whatever schema the user projected, that is what is returned [we drop identity partition columns not projected by the user]